# Data Manipulation with Pandas Library


`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

If you're curious:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself. After getting introduced, you can consult the full [`pandas` documentation](http://pandas.pydata.org/pandas-docs/stable/).

### Setting the working directory

Before loading the data, let's begin by setting the right working directory. In order to change the working directory, we use the `os` library

In [3]:
import os
os.getcwd()

'/Users/ghitaalami/Coding Projects/Berkeley Projects/Data-Science-with-VC-apps/Manipulation with pd'

### Loading CSV files

Now we can use the `pandas` library to load the data. Import `pandas` using the conventional abbreviation and call the `read_csv` method on your file's path name

In [4]:
import pandas as pd

# WHO = pd.read_csv("WHO.csv")
WHO = pd.read_csv("WHO.csv", encoding = "ISO-8859-1") ##check this encoding, also encoding = 'unicode_escape'

### The Dataframe

In [5]:
WHO

Country                 Region  Population  \
0                           Afghanistan  Eastern Mediterranean       29825   
1                               Albania                 Europe        3162   
2                               Algeria                 Africa       38482   
3                               Andorra                 Europe          78   
4                                Angola                 Africa       20821   
..                                  ...                    ...         ...   
189  Venezuela (Bolivarian Republic of)               Americas       29955   
190                            Viet Nam        Western Pacific       90796   
191                               Yemen  Eastern Mediterranean       23852   
192                              Zambia                 Africa       14075   
193                            Zimbabwe                 Africa       13724   

     Under15  Over60  FertilityRate  LifeExpectancy  ChildMortality  \
0      47.42    3.82           5.40              60            98.5   
1      21.33   14.93           1.75              74            16.7   
2      27.42    7.17           2.83              73            20.0   
3      15.20   22.86            NaN              82             3.2   
4      47.58    3.84           6.10              51           163.5   
..       ...     ...            ...             ...             ...   
189    28.84    9.17           2.44              75            15.3   
190    22.87    9.32           1.79              75            23.0   
191    40.72    4.54           4.35              64            60.0   
192    46.73    3.95           5.77              55            88.5   
193    40.24    5.68           3.64              54            89.8   

     CellularSubscribers  LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
0                  54.26           NaN   1140.0                          NaN   
1                  96.39           NaN   8820.0                          NaN   
2                  98.99           NaN   8310.0                         98.2   
3                  75.49           NaN      NaN                         78.4   
4                  48.38          70.1   5230.0                         93.1   
..                   ...           ...      ...                          ...   
189                97.78           NaN  12430.0                         94.7   
190               143.39          93.2   3250.0                          NaN   
191                47.05          63.9   2170.0                         85.5   
192                60.59          71.2   1490.0                         91.4   
193                72.13          92.2      NaN                          NaN   

     PrimarySchoolEnrollmentFemale  
0                              NaN  
1                              NaN  
2                             96.4  
3                             79.4  
4                             78.2  
..                             ...  
189                           95.1  
190                            NaN  
191                           70.5  
192                           93.9  
193                            NaN  

[194 rows x 13 columns]

In [6]:
# Structure of the data
WHO.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        194 non-null    object 
 1   Region                         194 non-null    object 
 2   Population                     194 non-null    int64  
 3   Under15                        194 non-null    float64
 4   Over60                         194 non-null    float64
 5   FertilityRate                  183 non-null    float64
 6   LifeExpectancy                 194 non-null    int64  
 7   ChildMortality                 194 non-null    float64
 8   CellularSubscribers            184 non-null    float64
 9   LiteracyRate                   103 non-null    float64
 10  GNI                            162 non-null    float64
 11  PrimarySchoolEnrollmentMale    101 non-null    float64
 12  PrimarySchoolEnrollmentFemale  101 non-null    flo

In [7]:
# Recent statistics from the World Health Organization (WHO)
# The variables are: 
# the name of the country
# the region the country is in
# the population in thousandsa
# the percentage of the population under 15 and over 60
# the fertility rate (average number of children per woman)
# the Life Expectancy in years
# the Child Mortality rate (the number of children who die by age 5 per 1000 births)
# the number of cellular subscribers per 100 population
# the literacy rate among adults aged >= 15
# the gross national income per capita
# the percentage of male children enrolled in primary school
# the percentage of female children enrolled in primary school

In [8]:
# Statistical summary of the data:
WHO.describe()

Population     Under15      Over60  FertilityRate  LifeExpectancy  \
count  1.940000e+02  194.000000  194.000000     183.000000      194.000000   
mean   3.635997e+04   28.732423   11.163660       2.940656       70.010309   
std    1.379031e+05   10.534573    7.149331       1.480984        9.259075   
min    1.000000e+00   13.120000    0.810000       1.260000       47.000000   
25%    1.695750e+03   18.717500    5.200000       1.835000       64.000000   
50%    7.790000e+03   28.650000    8.530000       2.400000       72.500000   
75%    2.453525e+04   37.752500   16.687500       3.905000       76.000000   
max    1.390000e+06   49.990000   31.920000       7.580000       83.000000   

       ChildMortality  CellularSubscribers  LiteracyRate           GNI  \
count      194.000000           184.000000    103.000000    162.000000   
mean        36.148969            93.641522     83.710680  13320.925926   
std         37.992935            41.400447     17.530645  15192.988650   
min          2.200000             2.570000     31.100000    340.000000   
25%          8.425000            63.567500     71.600000   2335.000000   
50%         18.600000            97.745000     91.800000   7870.000000   
75%         55.975000           120.805000     97.850000  17557.500000   
max        181.600000           196.410000     99.800000  86440.000000   

       PrimarySchoolEnrollmentMale  PrimarySchoolEnrollmentFemale  
count                   101.000000                     101.000000  
mean                     90.850495                      89.632673  
std                      11.017147                      12.817614  
min                      37.200000                      32.500000  
25%                      87.700000                      87.300000  
50%                      94.700000                      95.100000  
75%                      98.100000                      97.900000  
max                     100.000000                     100.000000

In [9]:
# Display a few data points at the "head" (start) of the dataset, i.e. the first few records
# By default, the first 5 records are shown; this can be overwritten by specificing the number of records in the paranthesis.
WHO.head(2) # try WHO.head(2)

Country                 Region  Population  Under15  Over60  \
0  Afghanistan  Eastern Mediterranean       29825    47.42    3.82   
1      Albania                 Europe        3162    21.33   14.93   

   FertilityRate  LifeExpectancy  ChildMortality  CellularSubscribers  \
0           5.40              60            98.5                54.26   
1           1.75              74            16.7                96.39   

   LiteracyRate     GNI  PrimarySchoolEnrollmentMale  \
0           NaN  1140.0                          NaN   
1           NaN  8820.0                          NaN   

   PrimarySchoolEnrollmentFemale  
0                            NaN  
1                            NaN

In [10]:
# Display the last few records.
WHO.tail()

Country                 Region  Population  \
189  Venezuela (Bolivarian Republic of)               Americas       29955   
190                            Viet Nam        Western Pacific       90796   
191                               Yemen  Eastern Mediterranean       23852   
192                              Zambia                 Africa       14075   
193                            Zimbabwe                 Africa       13724   

     Under15  Over60  FertilityRate  LifeExpectancy  ChildMortality  \
189    28.84    9.17           2.44              75            15.3   
190    22.87    9.32           1.79              75            23.0   
191    40.72    4.54           4.35              64            60.0   
192    46.73    3.95           5.77              55            88.5   
193    40.24    5.68           3.64              54            89.8   

     CellularSubscribers  LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
189                97.78           NaN  12430.0                         94.7   
190               143.39          93.2   3250.0                          NaN   
191                47.05          63.9   2170.0                         85.5   
192                60.59          71.2   1490.0                         91.4   
193                72.13          92.2      NaN                          NaN   

     PrimarySchoolEnrollmentFemale  
189                           95.1  
190                            NaN  
191                           70.5  
192                           93.9  
193                            NaN

### Subsets of data

In [11]:
# find the subset with only the countries in Europe
WHO_Europe = WHO[WHO['Region'] == 'Europe']
WHO_Europe.head()

Country  Region  Population  Under15  Over60  FertilityRate  \
1      Albania  Europe        3162    21.33   14.93           1.75   
3      Andorra  Europe          78    15.20   22.86            NaN   
7      Armenia  Europe        2969    20.34   14.06           1.74   
9      Austria  Europe        8464    14.51   23.52           1.44   
10  Azerbaijan  Europe        9309    22.25    8.24           1.96   

    LifeExpectancy  ChildMortality  CellularSubscribers  LiteracyRate  \
1               74            16.7                96.39           NaN   
3               82             3.2                75.49           NaN   
7               71            16.4               103.57          99.6   
9               81             4.0               154.78           NaN   
10              71            35.2               108.75           NaN   

        GNI  PrimarySchoolEnrollmentMale  PrimarySchoolEnrollmentFemale  
1    8820.0                          NaN                            NaN  
3       NaN                         78.4                           79.4  
7    6100.0                          NaN                            NaN  
9   42050.0                          NaN                            NaN  
10   8960.0                         85.3                           84.1

In [12]:
WHO_Europe.count()

Country                          53
Region                           53
Population                       53
Under15                          53
Over60                           53
FertilityRate                    50
LifeExpectancy                   53
ChildMortality                   53
CellularSubscribers              51
LiteracyRate                     26
GNI                              48
PrimarySchoolEnrollmentMale      38
PrimarySchoolEnrollmentFemale    38
dtype: int64

In [13]:
# Use compound boolean operators in the conditional expression
# Use & for and; | for or. 
WHO_AsiaEurope = WHO[(WHO['Region'] == 'Europe') | (WHO['Region'] == 'South-East Asia') | (WHO['Region'] == "Eastern Mediterranean")] 
WHO_AsiaEurope.head()

Country                 Region  Population  Under15  Over60  \
0  Afghanistan  Eastern Mediterranean       29825    47.42    3.82   
1      Albania                 Europe        3162    21.33   14.93   
3      Andorra                 Europe          78    15.20   22.86   
7      Armenia                 Europe        2969    20.34   14.06   
9      Austria                 Europe        8464    14.51   23.52   

   FertilityRate  LifeExpectancy  ChildMortality  CellularSubscribers  \
0           5.40              60            98.5                54.26   
1           1.75              74            16.7                96.39   
3            NaN              82             3.2                75.49   
7           1.74              71            16.4               103.57   
9           1.44              81             4.0               154.78   

   LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
0           NaN   1140.0                          NaN   
1           NaN   8820.0                          NaN   
3           NaN      NaN                         78.4   
7          99.6   6100.0                          NaN   
9           NaN  42050.0                          NaN   

   PrimarySchoolEnrollmentFemale  
0                            NaN  
1                            NaN  
3                           79.4  
7                            NaN  
9                            NaN

### Saving dataframe to CSV file

In [14]:
WHO_AsiaEurope.to_csv("WHO_AsiaEurope.csv")

Exercise 1.3

In [15]:
# Your turn:
# How many countries have population greater than 50 million? 
who_pop=WHO[WHO["Population"]>50000]
who_pop.count()[0]

25

### More Data Analysis

To access a variable in a data frame, you always have to link it to the data frame and call it using square brackets and pass it's name as a string.

In [18]:
# Now, run this.
WHO['LifeExpectancy']

0      60
1      74
2      73
3      82
4      51
       ..
189    75
190    75
191    64
192    55
193    54
Name: LifeExpectancy, Length: 194, dtype: int64

### Statistics

In [19]:
# Statistics of a variable
print((WHO['LifeExpectancy'].mean()))
print((WHO['LifeExpectancy'].max()))
print((WHO['LifeExpectancy'].min()))

70.01030927835052
83
47


In [20]:
# Standard deviation
WHO['LifeExpectancy'].std
WHO['LifeExpectancy'].describe()

count    194.000000
mean      70.010309
std        9.259075
min       47.000000
25%       64.000000
50%       72.500000
75%       76.000000
max       83.000000
Name: LifeExpectancy, dtype: float64

In [21]:
WHO['GNI'].describe()
# what's different here?

count      162.000000
mean     13320.925926
std      15192.988650
min        340.000000
25%       2335.000000
50%       7870.000000
75%      17557.500000
max      86440.000000
Name: GNI, dtype: float64

Here there are less instances (162) compared to the previous cell. This can be due to the presence of NaN values.

In [22]:
# Identify countries corresponding to max and min
idx_min = WHO['LifeExpectancy'].argmin()
print(WHO.iloc[idx_min]['Country'])

idx_max = WHO['LifeExpectancy'].argmax()
print(WHO['Country'][idx_max])

Sierra Leone
Japan


Exercise 1.4

In [23]:
# Exercise:
# What is the largest population value among all countries?
# Which country has the largest population?

idx_max_pop=WHO['Population'].argmax()
print(WHO['Population'][idx_max_pop])
print(WHO['Country'][idx_max_pop])


1390000
China


### Dealing with missing data

In [24]:
# Dealing with NA
# Try:
WHO['LiteracyRate'].head()

0     NaN
1     NaN
2     NaN
3     NaN
4    70.1
Name: LiteracyRate, dtype: float64

In [25]:
WHO.dropna(subset=['LiteracyRate'], inplace=True)
WHO['LiteracyRate'].head()
# Note: setting "inplace = True" will modify the original dataframe. 
# Alternatively, setting "inplace = False" will generate a new dataframe.

4     70.1
5     99.0
6     97.8
7     99.6
12    91.9
Name: LiteracyRate, dtype: float64

# References
- [1] Special thanks to the [EECS127 Fall 2019](https://inst.eecs.berkeley.edu/~ee127/fa19/) for providing a great starting point for Intro to Jupyter
- [2] D-lab intro to pandas
- [3] The official Python 3 language documentation. [Link](https://docs.python.org/3/).
- [4] The official numpy and scipy documentation. [Link](https://docs.scipy.org/doc/).


